performance-measure/src/main/java/org/sonarsource/performance/measure/PerformanceMeasure.java

```diff
@@ -0,0 +1,264 @@
+    globalDeactivation = value;
+  }
+
+  public static Builder reportBuilder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+
+    private boolean active = false;
+
+    @Nullable
+    private String performanceFile = null;
+
+    boolean appendMeasurementCost = false;
+
+    public Builder activate(boolean active) {
+      this.active = active;
+      return this;
+    }
+
+    public Builder appendMeasurementCost() {
+      this.appendMeasurementCost = true;
+      return this;
+    }
+
+    public Builder toFile(@Nullable String performanceFile) {
+      this.performanceFile = performanceFile;
+      return this;
+    }
+
+    public Duration start(String name) {
+      if (!active) {
+        return IgnoredDuration.INSTANCE;
+      }
+      setGlobalDeactivation(false);
+      Path performanceMeasureFile = null;
+      if (performanceFile != null && !performanceFile.isEmpty()) {
+        performanceMeasureFile = Paths.get(performanceFile.replace('\\', File.separatorChar).replace('/', File.separatorChar));
+      }
+      DurationMeasure parentMeasure = THREAD_LOCAL_CURRENT_MEASURE.get();
+      DurationMeasure currentMeasure = new DurationMeasure(name);
+      THREAD_LOCAL_CURRENT_MEASURE.set(currentMeasure);
+      return new RecordedDurationReport(parentMeasure, currentMeasure, performanceMeasureFile, appendMeasurementCost);
+    }
+
+  }
+
+  public static Duration start(Object object) {
+    if (globalDeactivation) {
+      return IgnoredDuration.INSTANCE;
+    }
+    return start(object.getClass().getSimpleName());
+  }
+
+  public static Duration start(String name) {
+    if (globalDeactivation) {
+      return IgnoredDuration.INSTANCE;
+    }
+    DurationMeasure parentMeasure = THREAD_LOCAL_CURRENT_MEASURE.get();
+    if (parentMeasure == null) {
+      return IgnoredDuration.INSTANCE;
+    }
+    DurationMeasure currentMeasure = parentMeasure.getOrCreateChild(name);
+    THREAD_LOCAL_CURRENT_MEASURE.set(currentMeasure);
+    return new RecordedDuration(parentMeasure, currentMeasure);
+  }
+
+  private static class RecordedDuration implements Duration {
+
+    @Nullable
+    protected final DurationMeasure parentMeasure;
+    protected final DurationMeasure measure;
+    private long startNanos;
+
+    public RecordedDuration(@Nullable DurationMeasure parentMeasure, DurationMeasure measure) {
+      this.parentMeasure = parentMeasure;
+      this.measure = measure;
+      this.startNanos = nanoTimeSupplier.get();
+    }
+
+    @Override
+    public void stop() {
+      if (startNanos != -1) {
+        measure.addCalls(1, nanoTimeSupplier.get() - startNanos);
+        startNanos = -1;
+        if (parentMeasure == null) {
+          THREAD_LOCAL_CURRENT_MEASURE.remove();
+        } else {
+          THREAD_LOCAL_CURRENT_MEASURE.set(parentMeasure);
+        }
+      }
+    }
+
+  }
+
+  private static class RecordedDurationReport extends RecordedDuration {
```

Comment:
I am not sure why `RecordedDurationReport` inherits from `RecordedDuration` when it could use an instance of `RecordedDuration` as an attribute ?